File: game/scores/high_score_renderer.py
import random
import pygame
import pygame.freetype

import global_state

class HighScoreRenderer:

    def __init__(self, scores, games):
        self.scores = scores
        self.score_font = pygame.freetype.SysFont("Arial", 28, bold=True)
        self.score_game = next(iter(games))
        self.games = games
        self.time_score = global_state.frame_time

    def set_random_score_game(self):
        if len(self.games) == 1:
            self.score_game = next(iter(self.games))
            return
        new_game = random.choice(list(self.games.keys()))
        while new_game == self.score_game:
            new_game = random.choice(list(self.games.keys()))
        self.score_game = new_game

    def render_outline(self, screen, text, xpos, ypos):
        text_surface_bg, _ = self.score_font.render(text, (0, 0, 0))
        text_surface_fg, _ = self.score_font.render(text, (255, 255, 255))
        screen.blit(text_surface_bg, (xpos - 1, ypos - 1))
        screen.blit(text_surface_bg, (xpos + 1, ypos - 1))
        screen.blit(text_surface_bg, (xpos - 1, ypos + 1))
        screen.blit(text_surface_bg, (xpos + 1, ypos + 1))
        screen.blit(text_surface_bg, (xpos - 2, ypos - 2))
        screen.blit(text_surface_bg, (xpos + 2, ypos - 2))
        screen.blit(text_surface_bg, (xpos - 2, ypos + 2))
        screen.blit(text_surface_bg, (xpos + 2, ypos + 2))
        screen.blit(text_surface_fg, (xpos, ypos))

    def render(self, screen):
        if global_state.frame_time - self.time_score > 4:
            self.set_random_score_game()
            self.time_score = global_state.frame_time

        self.render_outline(screen, "PUNTAJES " + self.games[self.score_game]["name"], 30, 30)

        top_scores = self.scores.get_top_scores(self.score_game, 5)
        for rank, (name, score) in enumerate(top_scores, start=1):
            self.render_outline(screen, f"{rank}. {name} - {score}", 30, 30 + rank * 30)

File: game-web/src/scores/highScoreRenderer.ts
/** Port of game/scores/high_score_renderer.py */

import { globalState } from '../globalState';
import type { Scores } from './scores';

export class HighScoreRenderer {
  private scores: Scores;
  private scoreGame: string;
  private games: Record<string, { name: string }>;
  private timeScore: number;

  constructor(scores: Scores, games: Record<string, { name: string }>) {
    this.scores = scores;
    this.games = games;
    this.scoreGame = Object.keys(games)[0];
    this.timeScore = globalState.frameTime;
  }

  setRandomScoreGame(): void {
    const keys = Object.keys(this.games);
    if (keys.length === 1) { this.scoreGame = keys[0]; return; }
    let newGame = keys[Math.floor(Math.random() * keys.length)];
    while (newGame === this.scoreGame) {
      newGame = keys[Math.floor(Math.random() * keys.length)];
    }
    this.scoreGame = newGame;
  }

  renderOutline(ctx: OffscreenCanvasRenderingContext2D, text: string, xpos: number, ypos: number): void {
    ctx.font = 'bold 28px Arial';
    const offsets = [-1, 1, -2, 2];
    ctx.fillStyle = '#000';
    for (const dx of offsets) {
      for (const dy of offsets) {
        ctx.fillText(text, xpos + dx, ypos + dy);
      }
    }
    ctx.fillStyle = '#fff';
    ctx.fillText(text, xpos, ypos);
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    if (globalState.frameTime - this.timeScore > 4) {
      this.setRandomScoreGame();
      this.timeScore = globalState.frameTime;
    }

    this.renderOutline(ctx, 'PUNTAJES ' + this.games[this.scoreGame].name, 30, 30);

    const topScores = this.scores.getTopScores(this.scoreGame, 5);
    topScores.forEach(([name, score], i) => {
      this.renderOutline(ctx, `${i + 1}. ${name} - ${score}`, 30, 30 + (i + 1) * 30);
    });
  }
}
